docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md

```diff
@@ -0,0 +1,265 @@
+Runtime Extension server.
+
+Please note that Cluster API only enforces request and response types as defined by a Runtime Hook version,
+and developers are fully responsible for all other elements of the design of a Runtime Extension implementation,
+including:
+
+- To choose which programming language to use; please note that for sake of this proposal Golang is the language
+  of choice, and we are not planning to test/provide tooling/libraries for other languages. Nevertheless, given that
+  we rely on Open API and plain HTTP(s) calls, other languages should just work but support will be provided at
+  best effort.
+- To choose if to have a dedicated HTTP server(s) for Runtime Extensions only or if to use the HTTP server for other
+  purposes as well (e.g. to serve a metric endpoint).
+
+In case the Runtime Extension is developed in Golang, the implementer can benefit from the following packages (exported by
+`sigs.k8s.io/cluster-api`) as shown in the example above:
+
+- `exp/runtime/hooks/api/v1alpha1` contains the Runtime Hook Golang API types, which are also used to generate the OpenAPI specification.
+- `exp/runtime/catalog` provides the `Catalog` object to register Runtime Hook definitions. The `Catalog` is then used by the `server` 
+  package to handle requests. `Catalog` is similar to the `runtime.Scheme` of the `k8s.io/apimachinery/pkg/runtime`
+  package just for Runtime Hooks.
+* `./exp/runtime/server`: The `server` package provides a `Server` object which makes it easy to implement
+  a Runtime Extension server. The `Server` will automatically handle tasks like Marshal/Unmarshal. A Runtime Extension 
+  developer only has to implement a strongly typed function that contains the actual extension implementation.
+
+## Guidelines
+
+While writing the actual code of the Runtime Extension a set of important guidelines must be considered:
+
+### Timeouts
+
+Runtime Extension processing adds to network request latency, they should run as quickly as possible
+(typically in milliseconds); Cluster Administrator will be allowed to configure how long the Cluster API Runtime
+should wait for a Runtime Extension to respond before treating the call as a failure (max 10s).
+
+### Availability
+
+Runtime Extension failure could result in errors in handling the workload clusters lifecycle, and so the implementation
+should be robust, have proper error handling, avoid panics, etc.. It will be allowed to set up
+failure policies preventing a Runtime Extension failure to have negative effects on the Cluster API Runtime, but
+this option can’t be used in all use cases (see [Error Management](#error-management)).
+
+### Blocking Hooks
+
+A Runtime Hook can be defined as "blocking", e.g. the `BeforeClusterUpgrade` hook allows a Runtime Extension
+to prevent the upgrade to start. A Runtime Extension registered for the `BeforeClusterUpgrade` hook can block 
+by retuning a non-zero `retryAfterSeconds` value. Following consideration apply:
+
```

Comment:
```suggestion
A Runtime Hook can be defined as "blocking" -  e.g. the `BeforeClusterUpgrade` hook allows a Runtime Extension
to prevent the upgrade process from starting. A Runtime Extension registered for the `BeforeClusterUpgrade` hook can block 
by retuning a non-zero `retryAfterSeconds` value. Following consideration apply:

```